src/runtime/chan.go

```diff
@@ -476,14 +487,18 @@ func chanSend(ch *channel, value unsafe.Pointer) {
 // The recieved value is copied into the value pointer.
 // Returns the comma-ok value.
 func chanRecv(ch *channel, value unsafe.Pointer) bool {
+	critical()
+
 	if rx, ok := ch.tryRecv(value); rx {
 		// value immediately available
 		chanDebug(ch)
+		exitCritical()
 		return ok
 	}
 
 	if ch == nil {
 		// A nil channel blocks forever. Do not schedule this goroutine again.
+		exitCritical()
 		deadlock()
 	}
 
```

Comment:
TODO: dont allocate in a critical section

Or better yet, don't allocate at all

Reply:
fix put into #1144